add forceErrorText to FormField & TextFormField.

hasanmhallak opened the pull request:
Introducing the `forceErrorText` property to both `FormField` and `TextFormField`. With this addition, we gain the capability to trigger an error state and provide an error message without invoking the `validator` method.

While the idea of making the `Validator` method work asynchronously may be appealing, it could introduce significant complexity to our current form field implementation. Additionally, this approach might not be suitable for all developers, as discussed by @justinmc  in this [comment](https://github.com/flutter/flutter/issues/56414#issuecomment-624960263).

This PR try to address this issue by adding `forceErrorText` property allowing us to force the error to the `FormField` or `TextFormField` at our own base making it possible to preform some async operations without the need for any hacks while keep the ability to check for errors if we call `formKey.currentState!.validate()`.

Here is an example:

<details> <summary>Code Example</summary>  

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(
    const MaterialApp(home: MyHomePage()),
  );
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({
    super.key,
  });

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final key = GlobalKey<FormState>();
  String? forcedErrorText;

  Future<void> handleValidation() async {
    // simulate some async work..
    await Future.delayed(const Duration(seconds: 3));

    setState(() {
      forcedErrorText = 'this username is not available.';
    });

    // wait for build to run and then check.
    //
    // this is not required though, as the error would already be showing.
    WidgetsBinding.instance.addPostFrameCallback((_) {
      print(key.currentState!.validate());
    });
  }

  @override
  Widget build(BuildContext context) {
    print('build');
    return Scaffold(
      floatingActionButton: FloatingActionButton(onPressed: handleValidation),
      body: Form(
        key: key,
        child: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Padding(
                padding: const EdgeInsets.symmetric(horizontal: 30),
                child: TextFormField(
                  forceErrorText: forcedErrorText,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```
</details>



Related to #9688 & #56414.

Happy to hear your thoughts on this.


## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

hasanmhallak committed:
add forceErrorText to FormField & TextFormField

hasanmhallak committed:
update forceErrorText to run on init & update tests

hasanmhallak committed:
Fix Linux docs_test

Fix unresolved doc reference errors

justinmc commented:
I agree that this kind of validation logic could be easier in Flutter, but I'm skeptical about adding this new parameter unless we're sure it's going to be worth it.

How is this better than the approach that I mentioned in the comment that you linked above?  I've update the dartpad here so that it runs on modern Flutter versions: https://dartpad.dartlang.org/?id=b8c96286ae3d35c47c202123197ff6a0

hasanmhallak committed:
Merge branch 'master' into master

justinmc commented:
Would you be able to post a full code example of the credit card case you mentioned, and/or the case of using bloc?

I'm actually on board with you that being able to simply provide error text to display unconditionally is a more simple and declarative pattern than what we have now.  However, I want to make sure we don't bloat the API into having multiple ways of doing things.  If what we really need is a bigger refactor or a breaking change then maybe we should take the time to plan that out instead.

I'm interested to look at those other code examples if you don't mind writing them, though!

hasanmhallak commented:
I'm not sure if I can provide the full code of the credit card case, but I can write some example to demonstrate more the cases that I have mention. Will be working on them and update the thread here.

hasanmhallak commented:
hey @justinmc, I've just created a small example, I tried not to use any library so you can able to check it on dart-pad (idk if dart-pad support using external libraries).

here's a link for the [example](https://dartpad.dev/?id=702430d19bd0da45cec76e02817bab05) on dart pad.

Now here, you may find it a bit confusing. But Riverpod/Bloc has a very nice way of filtering the state changes. But I couldn't simulate it without spending too much time.

```dart

 ValueListenableBuilder(
            valueListenable: service,
            builder: (context, value, child) {

              // this is the normal field without forcing any error.
             // it is here just to clean up the code a bit.
              final normalChild = _Field(
              controller: creditCardController,
              label: 'Credit Card',
              validator: Validators.creditCard,
            );

            return value.maybeWhen(
              orElse: () => normalChild,
              error: (type, message) {
                return switch (type) {
                 // filtring the error state related to cred card field to get
                // the error message if any.
                  ErrorType.creditCard => _Field(
                      controller: creditCardController,
                      label: 'Credit Card',
                      validator: Validators.creditCard,
                      // here I used the `errorText` property to simulate adding my suggestion to `FormField`
                      forceErrorText: message,
                    ),
                  _ => normalChild,
                      };
                    },
                  );
                },
              )
```

you can see in submit method that I saved a lot of boilerplate relating to caching the previous validated data
in order to run the validator again, and didn't be forced to use auto validation mode.

```dart
Future<void> submit() async {
    if (!_formKey.currentState!.validate()) return;

    // the approach I suggested won't let me register again when there is an error,
    // as the error would be forced to the [FormField] making the `validate` function return `false`.
    await service.register(
      username: usernameController.text,
      password: passwordController.text,
      creditCard: creditCardController.text,
    );
  }
```

The basic idea is to register a listener that will only be triggered if there is an error in an exact text field, so it will be rebuilt again with a forced error text.

this way we keep the logic for backed errors handling encapsulated in a (service, controller ..etc) and the ui will only be responsible for listening to any state change.

I think this way will give the developers more freedom on deciding how they would handle errors rather than sticking with one approach that can look like a work around.

Here is the link to the full flutter project on github [link](https://github.com/hasanmhallak/show-case-forcing-error-form-field)

hasanmhallak commented:
hey @justinmc had you got the time to review this?

hasanmhallak committed:
update docs

hasanmhallak committed:
Merge branch 'master' of github.com:hasanmhallak/flutter

hasanmhallak committed:
Merge branch 'master' into master

hasanmhallak committed:
Merge branch 'master' into master

hasanmhallak committed:
Merge branch 'master' into master

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -397,6 +398,23 @@ class FormField<T> extends StatefulWidget {
+  /// Note: Using [forceErrorText] directly to set the error state bypasses the regular
+  /// validation process performed by the [validator] function. This allows you to
+  /// handle error conditions or show specific error messages manually.
+  final String? forceErrorText;
```
Nit: What do you think of the name `errorTextOverride` instead? Looking through the docs, I think we use the term "override" in a similar way in a few other places, but I don't see "force" used like that.

justinmc replied:
Or we could even just call it `errorText`, which would match `InputDecoration.errorText`, since they both would do basically the same thing. However, I wonder if people would be confused about the interaction between this and `validator`.

hasanmhallak replied:
I think calling it `errorText` would indeed get people confused, and properly the same goes with `errorTextOverride`.

`errorTextOverride` would get people believe that it will simply override the text messages provided by the `validator`.

However, I'm with you on this one, as I don't really like the name `forceErrorText`, maybe if we make it `forcedErrorText`? the thing is the term `forced` will imply that we will force the error without any negotiation with `validator`.

Whats your take on this?

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -522,6 +562,13 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
   void _validate() {
+      if (widget.forceErrorText != null) {
+      _errorText.value = widget.forceErrorText;
+      // skip validating if error is forced.
```
Capital S here.

hasanmhallak replied:
Fix it

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -571,13 +618,34 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
+  void initState() {
+    super.initState();
+    if (widget.forceErrorText != null) {
+      // ignore: always_specify_types
```
I think do `Future<void>` to avoid this `ignore`.

hasanmhallak replied:
Fixed it

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -571,13 +618,34 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
+      // ignore: always_specify_types
+      Future.microtask(() {
+        setState(() {
+          _errorText.value = widget.forceErrorText;
```
Would this still work if you get rid of both the `Future.microtask` and the `setState`?

hasanmhallak replied:
I changed this approach entirely by marking the` _errorText` variable as `late` and setting it in initstate.

justinmc reviewed (commented):
@hasanmhallak Apologies for the long silence on this PR. I've been hoping we could get around to refactoring this instead of incrementally making fixes like this, but I think a refactor is still a ways off. Are you still interested in moving forward with this PR?

justinmc commented:
Somewhat related: https://github.com/flutter/flutter/issues/54104

hasanmhallak committed:
Merge branch 'flutter:master' into master

hasanmhallak committed:
Removed the `Future.microtask` in favor of a late variable

- Removed the `Future.microtask` in favor of a late variable.
- Fixed captial S in line 596

hasanmhallak committed:
Merge branch 'master' into master

hasanmhallak commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -651,14 +648,7 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
-        });
-      });
-    }
+    _errorText = RestorableStringN(widget.forceErrorText);
```
I removed the microtask in favor of a late variable to ensure that the `_errorText` Restorable String is registered in initState with the provided `forceErrorText` instead of registering it with `null` and then use setState to update it.

hasanmhallak commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -592,13 +591,11 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
+    if (widget.forceErrorText != null) {
       _errorText.value = widget.forceErrorText;
-      // skip validating if error is forced.
+      // Skip validating if error is forced.
```
Fixed capital S

hasanmhallak commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -519,7 +516,9 @@ class FormField<T> extends StatefulWidget {
-  final RestorableStringN _errorText = RestorableStringN(null);
+  // Marking it as late, so we can get to register it
+  // with the value provided by [forceErrorText].
+  late final RestorableStringN _errorText;
```
Marking `_errorText` as late.

justinmc commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -1060,7 +1060,133 @@ void main() {
+
+      expect(didCallValidator, isFalse);
+  });
+  
+  
```
Too much whitespace here and between the tests above.

justinmc commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -1060,7 +1060,133 @@ void main() {
+  });
+  
+  
+     testWidgets('Validator is nullified and error text behaves accordingly',
```
Indentation is weird here and in the tests above too.

justinmc commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -1125,5 +1251,6 @@ void main() {
     formKey.currentState!.validate();
     await tester.pump();
     expect(find.text('test_error'), findsNothing);
-  });
+    });
```
I think this indentation change is also wrong.

justinmc reviewed (commented):
The analyzer is failing, can you fix that? Looks like some trailing whitespace in form_test.dart.

hasanmhallak commented:
I formatted the form_test.dart using the dart formatter, this should resolve the problems.

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -465,6 +464,22 @@ class FormField<T> extends StatefulWidget {
+  /// The [forceErrorText] property is used to override the [InputDecoration.errorText]
+  /// value defined in the [InputDecoration] of the [TextFormField].
+  ///
+  /// Note: Using [forceErrorText] directly to set the error state bypasses the regular
```
See the styleguide on the use of "Note": https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-empty-prose

hasanmhallak replied:
I'll remove the note.

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -478,6 +493,15 @@ class FormField<T> extends StatefulWidget {
   /// height parent like [SizedBox], or set the [InputDecoration.helperText]
   /// parameter to a space.
+  ///
+  /// Note: Using [forceErrorText] will bypasses the regular
```
Another "Note": https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-empty-prose

hasanmhallak replied:
I'll be removing this too.

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -478,6 +493,15 @@ class FormField<T> extends StatefulWidget {
+  /// Note: Using [forceErrorText] will bypasses the regular
+  /// validation process performed by the [validator] function. This allows you to
+  /// handle error conditions or show specific error messages manually.
+  ///
+  /// See also:
+  ///
+  ///  * [forceErrorText], which may force the error state directly.
+  ///
```
I think this "See also" is unnecessary since it restates the previous paragraph. I would do one or the other.

hasanmhallak replied:
I removed this too.

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -533,16 +557,25 @@ class FormField<T> extends StatefulWidget {
+  /// [FormField.validator] callback is invoked, or If [FormField.forceErrorText] is set
+  /// directly to a non-null value.
+  ///
+  /// Note: When using [FormField.forceErrorText] directly to set the error state,
```
Another "Note" https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-empty-prose

hasanmhallak replied:
removed this too.

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -454,6 +452,7 @@ class FormField<T> extends StatefulWidget {
     super.key,
     required this.builder,
     this.onSaved,
+    this.forceErrorText,
     this.validator,
```
Is it true that both of these should not be used at the same time? Or what kind of cases do you expect like that?

hasanmhallak replied:
When using forceErrorText to display an error message, running the validator function to check the form's validity becomes redundant, as illustrated in the following code:
```dart
 void _validate() {
    if (widget.forceErrorText != null) {
      _errorText.value = widget.forceErrorText;
      // Skip validating if error is forced.
      return;
    }
    if (widget.validator != null) {
      _errorText.value = widget.validator!(_value);
    } else {
      _errorText.value = null;
    }
  }
```
The validator function provided by the user will not be called, and _validate will return false unless forceErrorText is set to null.

The initial issue was that InputDecoration has an errorText parameter, but it only affects the decoration, not the FormState. This results in a visual error state without affecting the form validation, causing _validate to return true.

To address this, I added a forceErrorText parameter to apply the error to both the decoration and the form state. This ensures that the form behaves as expected when an error is forced. The term "force" was chosen to indicate that the validator will not be called in this scenario.

justinmc replied:
Then I think it would make sense to disallow `validator` when `forceErrorText` is passed. I think you should add an `assert` in the initializers here to enforce that.

hasanmhallak replied:
The problem is that we need to use both. The main idea is that we cannot make the validator function run asynchronously because the form must always have a synchronous state (valid/invalid). Handling asynchronous validation errors within the form widget would introduce unnecessary complexity.

Instead, `forceErrorText` was introduced so the user can run the normal `validator` function to check for any potential errors, then perform any asynchronous computations. If an async validation error occurs, the user can simply set `forceErrorText` to display the error and make the `formState.validate()` function return false, so the form state is invalid. This approach keeps the form state management simple and synchronous.

Here, the return of the `validator` function becomes redundant as the user is knowingly forcing the error — hence the name forceErrorText — which is why there is no point in allowing the `validator` to be called. Adding an assertion to limit the use of only one of them would burden the user to alternate between providing the `forceErrorText` and a `validator` function. I think they can work together perfectly, as you don't have to call `formState.validate` to show the error of the forceErrorText. The moment the TextFormField is built with a forceErrorText, the error text will be displayed immediately.

The user is responsible for setting `forceErrorText` to `null` when the text value changes. This task should not be handled by the FormState widget, as it is beyond its intended scope.

justinmc replied:
Would you mind writing an example of that pattern you're describing? It could be something like https://github.com/flutter/flutter/blob/master/examples/api/lib/material/text_form_field/text_form_field.1.dart if you're up for it. Or if that's overkill then in a test or just paste the code here. I want to see how you intend to use this so I can have a better understanding of the API.

justinmc replied:
That example was great and helped me clearly see your point. I think you're right that it should be possible to pass both parameters at the same time.

We should probably document this behavior where we can, though. Maybe on the docs for each `forceErrorText` and `validator` parameter you should mention the other parameter and explain that `forceErrorText` will override `validator`.

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -465,6 +464,22 @@ class FormField<T> extends StatefulWidget {
+  /// will be set to the provided value, causing the form field to be considered
+  /// invalid and display the error message specified.
+  ///
+  /// The [forceErrorText] property is used to override the [InputDecoration.errorText]
+  /// value defined in the [InputDecoration] of the [TextFormField].
```
I wouldn't directly describe the parameter as it relates to another class in another library. Maybe just starting this with "for example" is ok... Or moving it to a "See also".

hasanmhallak replied:
I'll move it to a "See also".

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -562,7 +595,9 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
   ///
   ///  * [validate], which may update [errorText] and [hasError].
-  bool get isValid => widget.validator?.call(_value) == null;
+  ///
```
No need for an empty comment line here.

hasanmhallak replied:
okay.

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -562,7 +595,9 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
   ///  * [validate], which may update [errorText] and [hasError].
-  bool get isValid => widget.validator?.call(_value) == null;
+  ///
+  ///  * [FormField.forceErrorText], which may update [errorText] and [hasError].
```
"may update" => "also may update"

hasanmhallak replied:
fixed it.

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -562,7 +595,9 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
-  bool get isValid => widget.validator?.call(_value) == null;
+  ///
+  ///  * [FormField.forceErrorText], which may update [errorText] and [hasError].
+  bool get isValid => widget.forceErrorText != null || widget.validator?.call(_value) == null;
```
The logic seems incorrect or backwards to me here.

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -579,8 +614,12 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
+  /// Calls [FormField.validator] to set the [errorText] only if [FormField.forceErrorText] is null.
+  /// Returns true if there were no errors.
+  ///
+  /// Note: When using the [FormField.forceErrorText] property directly to set the
```
Another "Note" https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-empty-prose

hasanmhallak replied:
removed this too.

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -643,6 +687,20 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
+  @override
+  void didUpdateWidget(FormField<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.forceErrorText != oldWidget.forceErrorText) {
+      _errorText.value = widget.forceErrorText;
```
What if the validator is returning something nonnull at this point, do you still want to update this? Above I asked a question about whether or not it should be possible to use both at the same time. I think we need to answer that question first before deciding how this should work here.

justinmc replied:
That sounds reasonable if we do allow both parameters to be used at the same time (though I still want to see an example as requested below https://github.com/flutter/flutter/pull/132903/files#r1617904728).

justinmc reviewed (commented):
Can you undo the formatting changes to the unrelated parts of the test? It makes it hard to review.

hasanmhallak committed:
fix test_form indentation

hasanmhallak committed:
fix docs issues

hasanmhallak committed:
fix `isValid` and add test

hasanmhallak committed:
Merge branch 'master' into master

justinmc commented on packages/flutter/lib/src/material/text_form_field.dart:
```diff
@@ -105,6 +105,7 @@ class TextFormField extends FormField<String> {
     this.controller,
     String? initialValue,
     FocusNode? focusNode,
+    super.forceErrorText,
```
Similarly then, what happens if you pass both forceErrorText and InputDecorator.errorText?

hasanmhallak replied:
If both forceErrorText and InputDecorator.errorText are passed, the value of forceErrorText will override InputDecorator.errorText. This behavior remains unchanged and is handled at line 212.

justinmc replied:
Just to talk through an alternative, what if instead you pass InputDecorator.errorText to FormField.errorText?

hasanmhallak replied:
It's not practical to pass the errorText from InputDecorator to FormField's forceErrorText because FormField manages the logic to determine whether the form widget's state is valid or invalid, and it doesn't concern itself with the UI details. Since InputDecorator is responsible for displaying the form state in the UI, it is necessary to directly pass the errorText to InputDecorator to show errors. On the other hand, if we were to pass InputDecorator.errorText to FormField.forceErrorText, it would blur the lines between UI-specific implementation and form validation logic. InputDecorator should focus solely on rendering the decoration, and it shouldn't be tasked with forcing the Form into invalid state.

justinmc reviewed (commented):
Making sure we cover edge cases given overlapping parameters forceErrorText, validator, and InputDecorator.errorText.

justinmc commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -1060,8 +1060,178 @@ void main() {
+    expect(find.text(errorText), findsOneWidget);
+  });
+
+
```
Only 1 newline is needed here.

justinmc reviewed (commented):
I asked to see an example usage in one of my comments, I think that's the last main blocker. Sorry for all of the back and forth on this PR, I really appreciate your persistence. I want to take a look at exactly how you plan to use `validator` and `forceErrorText` together to make sure I understand everything.

hasanmhallak committed:
Merge remote-tracking branch 'upstream/master'

hasanmhallak committed:
remove extra empty line

hasanmhallak committed:
add text form field 2 example

hasanmhallak commented:
I added an example to illustrate how `validator` & `forceErrorText` can work together. I appreciate you taking the time to review this PR.

hasanmhallak committed:
link text form field 2 example to source file API doc comments

hasanmhallak committed:
add test for text form field example 2

hasanmhallak committed:
Merge remote-tracking branch 'upstream/master'

justinmc commented on examples/api/lib/material/text_form_field/text_form_field.2.dart:
```diff
@@ -0,0 +1,143 @@
+                      return 'Username must not contains any spaces';
+                    }
+                    if (int.tryParse(value[0]) != null) {
+                      return 'username must not start with a number';
```
Nit: Capitalize the U.

justinmc commented on examples/api/lib/material/text_form_field/text_form_field.2.dart:
```diff
@@ -0,0 +1,143 @@
+                      return 'username must not start with a number';
+                    }
+                    if (value.length <= 2) {
+                      return 'username should be at least 3 characters long';
```
Nit: Another U to capitalize.

justinmc commented on examples/api/lib/material/text_form_field/text_form_field.2.dart:
```diff
@@ -0,0 +1,143 @@
+                    return null;
+                  },
+                  onChanged: (String value) {
+                    if (forcedErrorText != null && !cachedServerErrors.containsKey(value)) {
+                      setState(() {
+                        forcedErrorText = null;
+                      });
+                    }
```
Nit: I think you can get rid of the forceErrorText state variable and just use cachedServerErrors. When you pass forceErrorText to TextFormField, you can lookup the username in cachedServerErrors at that point.

hasanmhallak replied:
I added a check for forcedErrorText to ensure it is not null, which helps save rebuild times when the user enters something that isn't cached. Without this check, setState would trigger a rebuild with each letter typed. However, I think this caching might confuse users. Therefore, I've updated the example to focus solely on forcedErrorText and how to utilize it effectively.

justinmc commented on examples/api/lib/material/text_form_field/text_form_field.2.dart:
```diff
@@ -0,0 +1,143 @@
+              mainAxisAlignment: MainAxisAlignment.center,
+              children: <Widget>[
+                TextFormField(
+                  forceErrorText: forcedErrorText,
```
Elsewhere I recommended getting rid of `forcedErrorText`, but if you keep it, I would drop the "d" so that its name matches the paramter `forceErrorText`.

justinmc commented on examples/api/lib/material/text_form_field/text_form_field.2.dart:
```diff
@@ -0,0 +1,143 @@
+  @override
+  void initState() {
+    super.initState();
+    controller = TextEditingController();
```
It's not necessary to instantiate this here, you can do it above on line 30.

justinmc commented on examples/api/lib/material/text_form_field/text_form_field.2.dart:
```diff
@@ -0,0 +1,143 @@
+                  decoration: const InputDecoration(
+                    hintText: 'Please write a username',
+                  ),
+                  validator: (String? value) {
```
Nit: This build method is getting long, maybe consider moving this validator into a static method above or something like that.

justinmc commented on examples/api/lib/material/text_form_field/text_form_field.2.dart:
```diff
@@ -0,0 +1,143 @@
+                else
+                  TextButton(
+                    child: const Text('Save'),
+                    onPressed: () async {
```
Another good candidate to move into a method above in order to make this more readable, if you think it would help.

justinmc commented on examples/api/test/material/text_form_field/text_form_field.2_test.dart:
```diff
@@ -0,0 +1,74 @@
+
+void main() {
+  group('TextFormFieldExample2 Widget Tests', () {
+
```
Nit: I would remove this newline.

justinmc commented on examples/api/test/material/text_form_field/text_form_field.2_test.dart:
```diff
@@ -0,0 +1,74 @@
+      final Finder textFormField = find.byType(TextFormField);
+      final Finder saveButton = find.byType(TextButton);
+
+      // Simulate entering a username already taken
```
Period at the end here.

justinmc commented on examples/api/test/material/text_form_field/text_form_field.2_test.dart:
```diff
@@ -0,0 +1,74 @@
+      await tester.enterText(textFormField, 'jack');
+      await tester.pump();
+      await tester.tap(saveButton);
+      await tester.pumpAndSettle(const Duration(seconds: 3)); // Wait for async call
```
Nit: Not sure, but you might be able to just use `pump` instead of `pumpAndSettle` here (keeping the duration).

justinmc replied:
Also, I would split this up. Pump one frame here,  then expect that the error message is not found (or no error message at all is found). Then `pump` the 3 second duration and continue the rest of the test as-is.

justinmc commented on packages/flutter/lib/src/material/text_form_field.dart:
```diff
@@ -81,6 +81,12 @@ export 'package:flutter/services.dart' show SmartDashesType, SmartQuotesType;
 /// {@end-tool}
 ///
+/// {@tool dartpad}
+/// This example shows how to force error text to the field and cache it.
```
I would add to the end: "after making an asynchronous call." or something like that.

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -465,6 +464,20 @@ class FormField<T> extends StatefulWidget {
+  ///
+  /// When the [forceErrorText] property is provided, the [FormFieldState.errorText]
+  /// will be set to the provided value, causing the form field to be considered
+  /// invalid and display the error message specified.
```
Nit: "and display" => "and to display"

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -465,6 +464,20 @@ class FormField<T> extends StatefulWidget {
+  /// When the [forceErrorText] property is provided, the [FormFieldState.errorText]
+  /// will be set to the provided value, causing the form field to be considered
+  /// invalid and display the error message specified.
+  ///
```
As mentioned above in another comment, I would mention `validator`. Something like this in a new paragraph maybe:

 > When [validator] is provided, [forceErrorText] will override any error that it returns. [validator] will not be called unless [forceErrorText] is null.

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -533,16 +546,22 @@ class FormField<T> extends StatefulWidget {
 class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
   late T? _value = widget.initialValue;
-  final RestorableStringN _errorText = RestorableStringN(null);
+  // Marking it as late, so we can get to register it
```
"we can get to register it" => "it can be registered"

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -465,6 +464,20 @@ class FormField<T> extends StatefulWidget {
+  /// See also:
+  ///
+  /// * [InputDecoration.errorText], which is used to display error messages in the text
+  /// field's decoration.
```
Also mention that forceErrorText takes priority over this as well.

justinmc reviewed (commented):
Thanks so much for the clear example, I agree that that is a totally valid use case and this is a good way to solve it.

Most of my comments are just nits and cleanup.

One question: Currently I think Form.validate will return `true` even when forceErrorText is nonnull, right? Just making sure that is intended. Maybe we should document it somewhere.

hasanmhallak committed:
refactor text form field example 2 to remove caching error and update test

hasanmhallak committed:
update text form field example 2 description

hasanmhallak committed:
update form docs

hasanmhallak commented:
I updated the provided example to remove the error caching mechanism, as I believe we should focus exclusively on demonstrating how to use `forceErrorText`.

In fact, Form.validate will return `false` if forceErrorText is not null. This is by design, as it indicates the form state is invalid, which is crucial to stopping the execution flow. For example:
```dart
// If [forceErrorText] is not null, `validate` will return false, stopping the execution
// and possibly preventing the end user from proceeding.
final bool isValid = formKey.currentState!.validate();
if (!isValid) {
  return;
}
```

hasanmhallak committed:
Merge remote-tracking branch 'upstream/master'

justinmc commented on examples/api/lib/material/text_form_field/text_form_field.2.dart:
```diff
@@ -0,0 +1,137 @@
+      return 'This field is required';
+    }
+    if (value.length != value.replaceAll(' ', '').length) {
+      return 'Username must not contains any spaces';
```
"contains" => "contain"

justinmc commented on examples/api/lib/material/text_form_field/text_form_field.2.dart:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  Future<void> onSave() async {
+    // Providing a default value in case if this was called on the
```
"in case if this" => "in case this"

justinmc commented on examples/api/lib/material/text_form_field/text_form_field.2.dart:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  return 'Username $username is already taken';
+}
```
I agree that this example looks better without the error caching. It more clearly illustrates the point.

justinmc commented on examples/api/test/material/text_form_field/text_form_field.2_test.dart:
```diff
@@ -0,0 +1,82 @@
+      await tester.tap(saveButton);
+      await tester.pump();
+      expect(find.text('Username jack is already taken'), findsNothing);
+      await tester.pump(const Duration(seconds: 3)); // Wait for async call.
```
Super nit: You could expose this duration in the example file and reference it here, so it's inherently clear why you're waiting without the need for the comment.

justinmc commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -465,6 +464,24 @@ class FormField<T> extends StatefulWidget {
+  /// will be set to the provided value, causing the form field to be considered
+  /// invalid and to display the error message specified.
+  ///
+  /// When [validator] is provided, [forceErrorText] will override any error that it
+  /// returns. [validator] will not be called unless [forceErrorText] is null.
```
Thanks for explaining `validator` here, this will be useful for people.

hasanmhallak replied:
No worries, Thank you for the reviewing this.

justinmc commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -1060,8 +1060,177 @@ void main() {
+    expect(didCallValidator, isFalse);
+  });
+
+  testWidgets('isValid returns false when forceErrorText is set and will change error display',
+  (WidgetTester tester) async {
```
I would format it like this:

```dart
  testWidgets(
    'isValid returns false when forceErrorText is set and will change error display',
    (WidgetTester tester) async {
      ...
```

justinmc commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -1060,8 +1060,177 @@ void main() {
+
+    await tester.pumpWidget(builder());
+
+    expect(fieldKey1.currentState!.isValid, isTrue);
```
Nit: I would also expect the value of `hasError` for `field1` here.

justinmc commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -1060,8 +1060,177 @@ void main() {
+    }
+
+    // Start off not autovalidating.
+    await tester.pumpWidget(builder(AutovalidateMode.disabled));
```
Can you also expect that forceErrorText is showing throughout this process?

justinmc replied:
And can you also expect that "this error should not..." does not appear?

justinmc commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -1060,8 +1060,177 @@ void main() {
+    expect(find.text(forceErrorText), findsOneWidget);
+  });
+
+  testWidgets('forceErrorText forces an error state only after setting it to a none-null value.', (WidgetTester tester) async {
```
"none-null" => "non-null"

justinmc commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -1060,8 +1060,177 @@ void main() {
+    }
+
+    await tester.pumpWidget(builder(AutovalidateMode.disabled, null));
+    await tester.pump();
```
Is this extra pump really needed?

hasanmhallak replied:
you're right, it's not needed.

justinmc commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -1060,8 +1060,177 @@ void main() {
-      (WidgetTester tester) async {
+
+  testWidgets('forceErrorText forces an error state when first init.', (WidgetTester tester) async {
+    final GlobalKey<FormState> formKey = GlobalKey<FormState>();
```
I think this is unused.

justinmc commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -1060,8 +1060,177 @@ void main() {
+    expect(fieldKey1.currentState!.isValid, isTrue);
+    expect(fieldKey2.currentState!.isValid, isFalse);
+    expect(fieldKey2.currentState!.hasError, isTrue);
+  });
```
I think this test could go in text_form_field_test.dart since it doesn't really have to do with Form directly.

justinmc replied:
Also in text_form_field_test.dart, could you add a test that shows what the behavior is when both InputDecoration.error and forceErrorText are set?

justinmc reviewed (approved):
LGTM with nits. I'll look for a secondary reviewer.

Can you add a test or expectation somewhere that `validate` returns false when forceErrorText is non-null, even when `validator` would otherwise return true?

hasanmhallak committed:
fix nits

hasanmhallak committed:
add tests to text form field

hasanmhallak committed:
Merge remote-tracking branch 'upstream/master'

hasanmhallak committed:
add test expect Validate to return false when forceErrorText is non-null

hasanmhallak commented:
Hey @justinmc, Thank you for reviewing this, I added the requested test, and fixed all the nits.

hasanmhallak committed:
fix failing test

Renzo-Olivares commented on examples/api/lib/material/text_form_field/text_form_field.2.dart:
```diff
@@ -0,0 +1,138 @@
+                  forceErrorText: forceErrorText,
+                  controller: controller,
+                  decoration: const InputDecoration(
+                    hintText: 'Please write a username',
```
What can a user expect here if they also define a `errorText` for their `InputDecoration`? I tried this behavior out on master (explicitly defining an InputDecoration.errorText and using `validate` with AutovalidateMode set to `onUserInteraction`) and the behavior was the following:

1. On initial build the widget will show the `errorText` defined in `InputDecoration`.
2. When typing and validate returns a non-null string it replaces the `InputDecoration.errorText`.
3. When typing and validate returns a null string it shows the explicit `InputDecoration.errorText` that I defined.

I think with `forceErrorText` set and `InputDecoration.errorText` set, the result should be that the explicit `errorText` is never shown unless `forceErrorText` is set to `null`. This would be good to test.

hasanmhallak replied:
> I think with forceErrorText set and InputDecoration.errorText set, the result should be that the explicit errorText is never shown unless forceErrorText is set to null. This would be good to test.

I already added a test to cover this behavior [here](https://github.com/flutter/flutter/pull/132903/files#diff-d604873acef188d0441304f2f30ac15c648c1268a6752efee225c834ace7609fR1767)

Renzo-Olivares commented on examples/api/lib/material/text_form_field/text_form_field.2.dart:
```diff
@@ -0,0 +1,138 @@
+  Widget build(BuildContext context) {
+    return Material(
+      child: Padding(
+        padding: const EdgeInsets.symmetric(horizontal: 24),
```
nit: use double literals for double constants `24` -> `24.0`, from style guide: https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#use-double-literals-for-double-constants

Renzo-Olivares commented on examples/api/lib/material/text_form_field/text_form_field.2.dart:
```diff
@@ -0,0 +1,138 @@
+                  validator: validator,
+                  onChanged: onChanged,
+                ),
+                const SizedBox(height: 40),
```
nit: see comment about double literals https://github.com/flutter/flutter/pull/132903/files#r1636910875

Renzo-Olivares commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -465,6 +464,24 @@ class FormField<T> extends StatefulWidget {
+  ///
+  /// See also:
+  ///
+  /// * [InputDecoration.errorText], which is used to display error messages in the text
```
What do you mean here by not effecting the field's state? I think explicitly setting `InputDecoration.errorText` changes the fields state to an error state right?

hasanmhallak replied:
`InputDecoration.errorText` are just for the `InputDecorator` widget and does not effect the `FormField` state. So calling `FormState.validate()` will return `true` even when we set the `InputDecoration.errorText` to a non-null value.

Renzo-Olivares replied:
I see what you mean, thank you for the clarification.

hasanmhallak replied:
of course.

hasanmhallak committed:
fix nits

hasanmhallak committed:
Merge remote-tracking branch 'upstream/master'

Renzo-Olivares commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -1060,8 +1060,161 @@ void main() {
 
-  testWidgets('Validator is nullified and error text behaves accordingly',
-      (WidgetTester tester) async {
+  testWidgets('forceErrorText forces an error state when first init.', (WidgetTester tester) async {
```
nit: Similar to other tests in this file, the test name should not end in a period. This is mainly to be consistent with other tests in this file.

Renzo-Olivares commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -1060,8 +1060,161 @@ void main() {
+      expect(isValid, isFalse);
+  });
+
+  testWidgets('forceErrorText forces an error state only after setting it to a non-null value.', (WidgetTester tester) async {
```
nit: See comment about test name, https://github.com/flutter/flutter/pull/132903/files#r1637108947

Renzo-Olivares commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -1060,8 +1060,161 @@ void main() {
+    expect(find.text(errorText), findsOne);
+  });
+
+  testWidgets('Validator will not be called if forceErrorText is provided.', (WidgetTester tester) async {
```
nit: See comment about test name, https://github.com/flutter/flutter/pull/132903/files#r1637108947

Renzo-Olivares reviewed (approved):
LGTM w/ small nits. Thank you for the contribution and patience @hasanmhallak!

hasanmhallak committed:
fix nits

hasanmhallak committed:
Merge remote-tracking branch 'upstream/master'

hasanmhallak commented:
@Renzo-Olivares Thank you for taking the time to review this. I fixed the nits you mentioned and updated the branch.

justinmc commented:
@hasanmhallak I think the golden bot message is incorrect, can you push a merge commit and see if that makes the flutter-golden check go green?

justinmc reviewed (approved):
Renewing my LGTM :+1:

hasanmhallak commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -1060,8 +1060,161 @@ void main() {
+  });
+
+  testWidgets(
+    'Validate returns false when forceErrorText is non-null even when Validator returns a null value',
```
of course.

hasanmhallak commented on packages/flutter/test/material/text_form_field_test.dart:
```diff
@@ -1714,4 +1714,84 @@ void main() {
+      expect(fieldKey2.currentState!.hasError, isTrue);
+  });
+
+  testWidgets('forceErrorText will override InputDecoration.error when both are provided', (WidgetTester tester) async {
```
of course!

hasanmhallak committed:
fix nits

hasanmhallak committed:
Merge remote-tracking branch 'upstream/master'

hasanmhallak commented:
@justinmc Thank you for reviewing this again, I really appreciate your time. I fixed the nits and merged with upstream.
